### PR TITLE
[scripts/debug_tools] Add CTU crash reduction script.

### PR DESCRIPTION
**Pull request by rnkovacs**
This script uses the existing debug tools to create a debug environment,
then preprocesses all source files in the failed ZIP, and runs C-Reduce
on them. As an optimization, it empties all source files that are found
not to contribute to the crash.

### CODE REVIEW

File: scripts/debug_tools/prepare_all_cmd_for_ctu.py [hunk 1]
```diff
@@ -17,24 +20,38 @@
...
+                              '%Y-%m-%d %H:%M')
+handler.setFormatter(formatter)
 
-        print("stdout:\n\n" + out.decode("utf-8"))
-        print("stderr:\n\n" + err.decode("utf-8"))
+LOG.setLevel(logging.INFO)
+LOG.addHandler(handler)
 
+
+def execute(cmd, cwd=None, env=None):
```
Comment (rnkovacs) [inline scripts/debug_tools/prepare_all_cmd_for_ctu.py]: Having `cwd` and `env` here is useful. The output format change is not necessary though, just made my output more consistent with CodeChecker's output. Are we relying on the previous format somewhere? Should I revert the changes?

File: scripts/debug_tools/reduce_ctu_crash.py [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+    test = assemble_test(opts.analyzed_file, opts, opts.analyze_std)
+    run_creduce(opts.analyzed_file, test, opts.jobs, counter)
+
+
+def get_cmd_after_preproc(old_cmd, file_dir_path):
+    """
+    Create a new compilation command for a preprocessed and relocated
+    source file.
+    """
+    old_cmd = __remove_dependency_generation(old_cmd.split(" "))
```
Comment (gyorb) [inline scripts/debug_tools/reduce_ctu_crash.py]: using [shlex.split](https://docs.python.org/3/library/shlex.html) is generally a better idea to split up compilation commands there might be arguments with space in them like `-Dvar="my value"` which would be split wrong..

File: scripts/debug_tools/reduce_ctu_crash.py [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+    LOG.info("Updated compile command: %s", new_cmd)
+    return new_cmd
+
+
+def get_new_analyzer_cmd(clang, old_cmd, output_dir, triple_arch):
+    """
+    Create a new CTU analyze command after files have been relocated
+    and modified for debugging.
+    """
+    old_cmd = old_cmd.split(" ")
```
Comment (gyorb) [inline scripts/debug_tools/reduce_ctu_crash.py]: using `shlex.split` here too?

File: scripts/debug_tools/reduce_ctu_crash.py [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+
+def parse_arguments():
+    parser = argparse.ArgumentParser(
+        description='Reduces reproduction files for CTU assertions '
+        'based on the repro zip file generated by CodeChecker. '
+        'This script requires CodeChecker to be added to PATH '
+        'and its environment sourced. You can specify which Clang to use '
+        '(it is taken from PATH by default).')
+
+    # Required arguments.
```
Comment (gyorb) [inline scripts/debug_tools/reduce_ctu_crash.py]: you can set it explicitly that this argument is [required](https://docs.python.org/3/library/argparse.html#required)

File: scripts/debug_tools/reduce_ctu_crash.py [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+    """
+    Run C-Reduce on the specified file with the given interestingness test.
+    """
+    if counter < 1:
+        failed = try_empty(test_script, reduced_file)
+        if not failed:
+            return
+
+    # FIXME: Is there a need to specify custom creduce commands?
+    # Maybe optinally read it from a file?
```
Comment (gyorb) [inline scripts/debug_tools/reduce_ctu_crash.py]:
```suggestion
    # Maybe optionally read it from a file?
```

File: scripts/debug_tools/prepare_all_cmd_for_ctu.py (hunk 1 above)
Comment (martong) [reply scripts/debug_tools/prepare_all_cmd_for_ctu.py]: It is okay like this, we don't depend on the output anywhere.

**Comment by rnkovacs** [comment]
Thank you for your reviews! I'll add the docs stuff as soon as I can.

**Comment by csordasmarton** [comment]
Closing this pull request as there have been no activity on it in a long time.
